rust/src/ketama.rs

```diff
@@ -72,6 +85,55 @@ fn create_server_ranking(keys: &[&str]) -> Vec<ServerRank> {
     }
 
     ranking.sort_by_key(|rank| rank.value);
+}
+
+fn ketama_hash(md5_hash: &[u8], alignment: usize) -> u32 {
+    u32::from_be_bytes([
+        md5_hash[3 + alignment * 4],
+        md5_hash[2 + alignment * 4],
+        md5_hash[1 + alignment * 4],
+        md5_hash[alignment * 4],
+    ])
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::ketama::*;
+    use md5::{Digest, Md5};
 
-    ranking
+    #[test]
+    fn it_produces_correct_ketama_hashes() {
+        let pool = KetamaPool::new(&["server1"]);
+        assert_eq!(pool.ranking.len(), 1 * POINTS_PER_SERVER * POINTS_PER_HASH);
+
+        let hash = Md5::digest("server1-8");
+        /*
+        Results taken from twemproxy tests:
+
+        expect_same_uint32_t(3853726576U, ketama_hash("server1-8", strlen("server1-8"), 0), "should have expected ketama_hash for server1-8 index 0");
+        expect_same_uint32_t(2667054752U, ketama_hash("server1-8", strlen("server1-8"), 3), "should have expected ketama_hash for server1-8 index 3");
+        */
+        assert_eq!(3853726576, ketama_hash(hash.as_slice(), 0));
+        assert_eq!(2667054752, ketama_hash(hash.as_slice(), 3));
+    }
+
+    #[test]
+    fn it_can_add_and_remove_servers() {
+        let mut pool = KetamaPool::new(&["server1"]);
+        pool.add_node("server2");
+        assert_eq!(2 * POINTS_PER_SERVER * POINTS_PER_HASH, pool.ranking.len());
```

Comment:
these assertions here and above are testing implementation details, and I don’t think it really adds any value?

What you are not testing is one of the properties of "consistent hashing":
- that after adding a node, an existing key has a X% chance to be assigned the same node.
- that after removing the added node, all existing keys retain their previous node assignment.